File: Modelos Java/MatrizJava/src/main/java/com/mycompany/matrizjava/Modelo/Modelo.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package com.mycompany.matrizjava.Modelo;

/**
 *
 * @author 57320
 */
import java.util.ArrayList;
import java.util.Random;

public class Modelo {
    public ArrayList<ArrayList<Politico>> matrizPoliticos = new ArrayList<>();
    public int k, m;

    public void generarPoliticos(int n) {
        matrizPoliticos.clear();
        Random rand = new Random();
        k = (int) Math.ceil(Math.sqrt(n));
        m = (int) Math.ceil((double) n / k);
        int total = k * m;

        ArrayList<Politico> politicos = new ArrayList<>();

        for (int i = 0; i < n; ++i) {
            Politico p = new Politico();
            p.id = i + 1;
            p.dineroRobado = rand.nextInt(1000000) + 100;
            p.edad = rand.nextInt(50) + 30;
            p.fechaNacimiento = String.valueOf(2024 - p.edad);
            politicos.add(p);
        }

        while (politicos.size() < total) {
            Politico p = new Politico();
            p.id = -1;
            p.dineroRobado = -1;
            p.edad = -1;
            p.fechaNacimiento = "X";
            politicos.add(p);
        }

        int index = 0;
        for (int i = 0; i < k; i++) {
            ArrayList<Politico> fila = new ArrayList<>();
            for (int j = 0; j < m; j++) {
                fila.add(politicos.get(index++));
            }
            matrizPoliticos.add(fila);
        }
    }

    public void ordenarPorEdad(int algoritmo) {
        for (ArrayList<Politico> fila : matrizPoliticos) {
            if (algoritmo == 1)
                bubbleSort(fila, true);
            else
                selectionSort(fila, true);
        }
    }

    public void ordenarPorDineroYEdad(int algoritmo) {
        for (ArrayList<Politico> fila : matrizPoliticos) {
            if (algoritmo == 1)
                bubbleSort(fila, false);
            else
                selectionSort(fila, false);
        }

        for (int j = 0; j < m; j++) {
            for (int i = 0; i < k - 1; i++) {
                int minIndex = i;
                for (int l = i + 1; l < k; l++) {
                    if (matrizPoliticos.get(l).get(j).edad < matrizPoliticos.get(minIndex).get(j).edad) {
                        minIndex = l;
                    }
                }
                Politico temp = matrizPoliticos.get(i).get(j);
                matrizPoliticos.get(i).set(j, matrizPoliticos.get(minIndex).get(j));
                matrizPoliticos.get(minIndex).set(j, temp);
            }
        }
    }

    private void bubbleSort(ArrayList<Politico> fila, boolean ordenarPorEdad) {
        for (int i = 0; i < fila.size() - 1; i++) {
            for (int j = 0; j < fila.size() - i - 1; j++) {
                if ((ordenarPorEdad && fila.get(j).edad > fila.get(j + 1).edad) ||
                    (!ordenarPorEdad && fila.get(j).dineroRobado > fila.get(j + 1).dineroRobado)) {
                    Politico temp = fila.get(j);
                    fila.set(j, fila.get(j + 1));
                    fila.set(j + 1, temp);
                }
            }
        }
    }

    private void selectionSort(ArrayList<Politico> fila, boolean ordenarPorEdad) {
        for (int i = 0; i < fila.size() - 1; i++) {
            int minIndex = i;
            for (int j = i + 1; j < fila.size(); j++) {
                if ((ordenarPorEdad && fila.get(j).edad < fila.get(minIndex).edad) ||
                    (!ordenarPorEdad && fila.get(j).dineroRobado < fila.get(minIndex).dineroRobado)) {
                    minIndex = j;
                }
            }
            Politico temp = fila.get(i);
            fila.set(i, fila.get(minIndex));
            fila.set(minIndex, temp);
        }
    }
}

File: Modelos Python/MatrizPython/modelo/modelo.py
import random
import math
from datetime import datetime

class Politico:
    def __init__(self, id, dinero_robado, fecha_nacimiento, edad):
        self.id = id
        self.dinero_robado = dinero_robado
        self.fecha_nacimiento = fecha_nacimiento
        self.edad = edad

class Modelo:
    def __init__(self):
        self.matriz_politicos = []
        self.k = 0
        self.m = 0

    def generar_politicos(self, n):
        self.k = math.ceil(math.sqrt(n))
        self.m = math.ceil(n / self.k)
        total = self.k * self.m
        politicos = []

        for i in range(n):
            edad = random.randint(30, 79)
            anio_nacimiento = datetime.now().year - edad
            p = Politico(
                id=i + 1,
                dinero_robado=random.randint(100, 999999),
                fecha_nacimiento=str(anio_nacimiento),
                edad=edad
            )
            politicos.append(p)

        while len(politicos) < total:
            politicos.append(Politico(-1, -1, "X", -1))

        self.matriz_politicos = []
        index = 0
        for i in range(self.k):
            fila = []
            for j in range(self.m):
                fila.append(politicos[index])
                index += 1
            self.matriz_politicos.append(fila)

    def ordenar_por_edad(self, usar_bubble_sort):
        for fila in self.matriz_politicos:
            if usar_bubble_sort:
                self.bubble_sort(fila, por_edad=True)
            else:
                self.selection_sort(fila, por_edad=True)

    def ordenar_por_dinero_y_edad(self, usar_bubble_sort):
        for fila in self.matriz_politicos:
            if usar_bubble_sort:
                self.bubble_sort(fila, por_edad=False)
            else:
                self.selection_sort(fila, por_edad=False)

        for j in range(self.m):
            for i in range(self.k - 1):
                min_index = i
                for l in range(i + 1, self.k):
                    if self.matriz_politicos[l][j].edad < self.matriz_politicos[min_index][j].edad:
                        min_index = l
                self.matriz_politicos[i][j], self.matriz_politicos[min_index][j] =                     self.matriz_politicos[min_index][j], self.matriz_politicos[i][j]

    def bubble_sort(self, fila, por_edad):
        for i in range(len(fila) - 1):
            for j in range(len(fila) - i - 1):
                if ((por_edad and fila[j].edad > fila[j + 1].edad) or
                    (not por_edad and fila[j].dinero_robado > fila[j + 1].dinero_robado)):
                    fila[j], fila[j + 1] = fila[j + 1], fila[j]

    def selection_sort(self, fila, por_edad):
        for i in range(len(fila) - 1):
            min_index = i
            for j in range(i + 1, len(fila)):
                if ((por_edad and fila[j].edad < fila[min_index].edad) or
                    (not por_edad and fila[j].dinero_robado < fila[min_index].dinero_robado)):
                    min_index = j
            fila[i], fila[min_index] = fila[min_index], fila[i]
